Jetpack E2E: Paywall block.

worldomonation opened the pull request:
Related to: https://github.com/Automattic/wp-calypso/issues/80730.

## Proposed Changes

This PR adds testing flow for the Paywall block.

Key flows:
- add block
- insert pre and post-paywall content.
- verify as unrelated user, the functionality of the paywall block.

## Testing Instructions

Ensure the following build configurations are passing:
  - [x] Calypso E2E (desktop)
  - [x] Jetpack E2E Simple (mobile)
  - [x] Jetpack E2E Simple (desktop)
  - [x] Jetpack E2E AT (desktop)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

worldomonation committed:
Initial round of work

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/paywall.ts
- implement block flow.

test/e2e/specs/blocks/blocks__jetpack-earn.ts
- implement Paywall Block flow.

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/donations-form.ts
- add a step to escape the modal

packages/calypso-e2e/src/lib/blocks/block-flows/paywall.ts
- finish the POM

packages/calypso-e2e/src/lib/pages/editor-page.ts
- add ability to specify timeout when publishing.

test/e2e/specs/blocks/blocks__jetpack-earn.ts
- reorder spec

test/e2e/specs/blocks/shared/block-smoke-testing.ts
- add timeout when visiting the published page

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/donations-form.ts
- fix typo.

packages/calypso-e2e/src/lib/blocks/block-flows/paywall.ts
- add comment.

test/e2e/specs/blocks/blocks__jetpack-earn.ts
- do not run for private sites.

worldomonation commented on packages/calypso-e2e/src/lib/blocks/block-flows/paywall.ts:
```diff
@@ -0,0 +1,88 @@
+			.waitFor();
+
+		// Unrelated user sees a paywall.
+		const browser = context.page.context().browser();
+
+		if ( ! browser ) {
+			throw new Error( `Failed to retrieve browser instance.` );
+		}
+		const newPage = await ( await browser.newContext() ).newPage();
+		await newPage.goto( context.page.url(), { waitUntil: 'domcontentloaded' } );
+
+		await newPage.getByRole( 'main' ).getByText( this.configurationData.prePaywallText ).waitFor();
+		await newPage
+			.getByRole( 'main' )
+			.getByText( this.configurationData.postPaywallText )
+			.waitFor( { state: 'detached' } );
```
This is kind of "hacky" in the sense that it's breaking the encapsulation of concerns imposed by the blockflow design pattern. 

That said, limiting the check to the publishing user seemed pointless. 

If you (@dpasque) think this is better off being a standalone test, I'm totally up for a quick rewrite.

dpasque replied:
Hmm yeah, this one is tough, because it seems to live right between the two mental models (block smoke testing vs a full dedicated flow.)

I think this is okay to keep as a block smoke test to start! We can always move around if needed.

One non-blocking suggestion: I think this might read a little cleaner (and help in future cases) if we actually just extend the `PublishedPostContext` to either have a pointer to `browser` or maybe even just a `newPage` function on it that we then provide in the `block-smoke-testing.ts` file. I feel like the `block-smoke-testing.ts` file should take care of getting the actual dependencies and the smoke test flows should only be concerned with using them. That conceptually makes sense to me! WDYT?

worldomonation replied:
Done! Added the `browser` object to the published context.

worldomonation commented on packages/calypso-e2e/src/lib/pages/editor-page.ts:
```diff
@@ -610,7 +610,10 @@ export class EditorPage {
 	 * @returns {URL} Published article's URL.
 	 */
-	async publish( { visit = false }: { visit?: boolean } = {} ): Promise< URL > {
+	async publish( {
+		visit = false,
+		timeout,
+	}: { visit?: boolean; timeout?: number } = {} ): Promise< URL > {
```
Not strictly necessary, but useful change to have so callers can override the default timeout. This is really useful for the Donations block on AT.

worldomonation committed:
test/e2e/specs/blocks/blocks__jetpack-earn.ts
- fix lint error

dpasque commented on test/e2e/specs/blocks/blocks__jetpack-earn.ts:
```diff
@@ -55,4 +56,18 @@ if (
 	blockFlows.push( new AdFlow( {} ) );
 }
 
+// Private sites also do not apply to Private sites.
```
Haha, I don't think this tautology is intentional. 🤣  Do you mean "paywall blocks" do not apply to private sites?

worldomonation replied:
Good catch and lol @ tautology

dpasque reviewed (approved):
LGTM! Two small suggestions 😄

worldomonation committed:
test/e2e/specs/blocks/blocks__jetpack-earn.ts
- fix tautology in comment

worldomonation committed:
test/e2e/specs/blocks/shared/block-smoke-testing.ts
packages/calypso-e2e/src/lib/blocks/block-flows/types.ts
- add new property 'browser' to the contexts object.

packages/calypso-e2e/src/lib/blocks/block-flows/paywall.ts
- use the passed in browser context.

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/paywall.ts
- set the noSearch flag by default
